[Bug]: Panic of jaeger collector

### What happened?

I have a panic of the jaeger-collector

### Steps to reproduce

1. Version of jaeger-collector : {"gitCommit":"fdc3ae5d914daa5c097bf18afbfa62e1c0d4de97","gitVersion":"v1.37.0","buildDate":"2022-08-03T17:38:30Z"} . In my previous version (1.26.0), I also reproduce the bug 
2. Have a configuration with both Kafka with protobuf encoding (no issue with json) + --collector.tags options
3. It is a random bug (or at least I don't identify precisely the faulting span) 
4. At 1500 spans/s, wait less than 30 minutes (sometimes 

### Expected behavior

No panic 

### Relevant log output

```shell
panic: runtime error: slice bounds out of range [-6:]
goroutine 123 [running]:
github.com/jaegertracing/jaeger/model.(*Span).MarshalToSizedBuffer(0xc000a2aff0, {0xc000a858c0, 0x29e, 0x29e})
github.com/jaegertracing/jaeger/model/model.pb.go:1246 +0x8c5
github.com/jaegertracing/jaeger/model.(*Span).XXX_Marshal(0xc000433620?, {0xc000a858c0, 0xc000a2aff0?, 0x29e}, 0x90?)
github.com/jaegertracing/jaeger/model/model.pb.go:377 +0x37
github.com/gogo/protobuf/proto.Marshal({0x18e13c0, 0xc000a2aff0})
github.com/gogo/protobuf@v1.3.2/proto/table_marshal.go:2940 +0x144
github.com/jaegertracing/jaeger/plugin/storage/kafka.(*protobufMarshaller).Marshal(0xc00160e2e0?, 0xc000619a30?)
github.com/jaegertracing/jaeger/plugin/storage/kafka/marshaller.go:39 +0x25
github.com/jaegertracing/jaeger/plugin/storage/kafka.(*SpanWriter).WriteSpan(0xc00073af00, {0xc0017d84b0?, 0x0?}, 0xc000a2aff0)
github.com/jaegertracing/jaeger/plugin/storage/kafka/writer.go:77 +0x39
github.com/jaegertracing/jaeger/storage/spanstore.(*CompositeWriter).WriteSpan(0x18e5698?, {0x18e5708, 0xc0017d84b0}, 0x18d3e10?)
github.com/jaegertracing/jaeger/storage/spanstore/composite.go:41 +0x11e
github.com/jaegertracing/jaeger/cmd/collector/app.(*spanProcessor).saveSpan(0xc00039a510, 0xc000a2aff0, {0x0, 0x0})
github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:160 +0xfe
github.com/jaegertracing/jaeger/cmd/collector/app.ChainedProcessSpan.func1(0xc0002aca80?, {0x0, 0x0})
github.com/jaegertracing/jaeger/cmd/collector/app/model_consumer.go:35 +0x4b
github.com/jaegertracing/jaeger/cmd/collector/app.(*spanProcessor).processItemFromQueue(0xc00039a510, 0xc0012f9c80)
github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:191 +0x57
github.com/jaegertracing/jaeger/cmd/collector/app.NewSpanProcessor.func1({0x133d440?, 0xc0012f9c80?})
github.com/jaegertracing/jaeger/cmd/collector/app/span_processor.go:78 +0x36
github.com/jaegertracing/jaeger/pkg/queue.ConsumerFunc.Consume(0xc000619fa8?, {0x133d440?, 0xc0012f9c80?})
github.com/jaegertracing/jaeger/pkg/queue/bounded_queue.go:105 +0x2c
github.com/jaegertracing/jaeger/pkg/queue.(*BoundedQueue).StartConsumersWithFactory.func1()
github.com/jaegertracing/jaeger/pkg/queue/bounded_queue.go:84 +0xd6
created by github.com/jaegertracing/jaeger/pkg/queue.(*BoundedQueue).StartConsumersWithFactory
github.com/jaegertracing/jaeger/pkg/queue/bounded_queue.go:74 +0x53
```


### Screenshot

_No response_

### Additional context

* The collector tags is important 
  * Without collector.tags, it doesn't panic
  * With collector.tags, it panics 

I dig a bit in the code (by adding some debug) and it seems that it could be linked to https://github.com/jaegertracing/jaeger/blob/ddca3c84c3731025163d5b5bc4b1b648bc904fbf/cmd/collector/app/span_processor.go#L213 as the sort seems to introduce a corruption in the data structure . 

For tests, I recompile without the typedTags.Sort() and I hadn't have panic anymore

### Jaeger backend version

1.37.0

### SDK

_No response_

### Pipeline

OTEL Shim => OTEL SDK => Jaeger Agent => Jaeger collector => Kafka (protobuf)

### Stogage backend

ElasticSearch 7 + Kafka (protobuf)

### Operating system

Linux

### Deployment model

binaries on KVM VMs 

### Deployment configs

```shell
export SPAN_STORAGE_TYPE=elasticsearch,kafka
export ES_USERNAME
export ES_PASSWORD
export KAFKA_PRODUCER_PLAINTEXT_USERNAME
export KAFKA_PRODUCER_PLAINTEXT_PASSWORD
/home/cloud-user/local/bin/jaeger-collector \
    --sampling.strategies-file=$HOME/local/etc/jaeger/sampling_strategies.json \
    --collector.queue-size=100000 \
    --es.server-urls=$ES_URLS \
    --es.num-replicas=1\
    --es.bulk.workers=10\
    --es.tags-as-fields.all\
    --kafka.producer.brokers $KAFKA_BROKERS\
    --kafka.producer.encoding protobuf\
    --kafka.producer.topic jaeger-spans-protobuf\
    --kafka.producer.authentication plaintext\
    --kafka.producer.plaintext.mechanism SCRAM-SHA-512
    --collector.tags="collector_name=$COLLECTOR_NAME,collector_az=$AZ,collector_region=$REGION"
```

--- a/cmd/collector/app/span_processor.go
+++ b/cmd/collector/app/span_processor.go
@@ -177,6 +177,16 @@ func (sp *spanProcessor) ProcessSpans(mSpans []*model.Span, options processor.Sp
 	sp.preProcessSpans(mSpans, options.Tenant)
 	sp.metrics.BatchSize.Update(int64(len(mSpans)))
 	retMe := make([]bool, len(mSpans))
+
+	// Note: this is not the ideal place to do this because collector tags are added to Process.Tags,
+	// and Process can be shared between different spans in the batch, but we no longer know that,
+	// the relation is lost upstream and it's impossible in Go to dedupe pointers. But at least here
+	// we have a single thread updating all spans that may share the same Process, before concurrency
+	// kicks in.
+	for _, span := range mSpans {
+		sp.addCollectorTags(span)
+	}
+
 	for i, mSpan := range mSpans {
 		ok := sp.enqueueSpan(mSpan, options.SpanFormat, options.InboundTransport, options.Tenant)
 		if !ok && sp.reportBusy {
@@ -213,6 +223,8 @@ func (sp *spanProcessor) addCollectorTags(span *model.Span) {
 	typedTags.Sort()
 }
 
+// Note: spans may share the Process object, so no changes should be made to Process
+// in this function as it may cause race conditions.
 func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat processor.SpanFormat, transport processor.InboundTransport, tenant string) bool {
 	spanCounts := sp.metrics.GetCountsForFormat(originalFormat, transport)
 	spanCounts.ReceivedBySvc.ReportServiceNameForSpan(span)
@@ -224,9 +236,6 @@ func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat processor.
 
 	// add format tag
 	span.Tags = append(span.Tags, model.String("internal.span.format", string(originalFormat)))
-
-	// append the collector tags
-	sp.addCollectorTags(span)
 
 	item := &queueItem{
 		queuedTime: time.Now(),

